src/state_machine/transitions.cpp

```diff
@@ -0,0 +1,177 @@
+ *    You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "state_machine/transitions.hpp"
+
+namespace hyped {
+
+namespace state_machine {
+
+//--------------------------------------------------------------------------------------
+// Emergency
+//--------------------------------------------------------------------------------------
+
+/*
+ * @brief   Local function that determines whether or not there is an emergency.
+ */
+inline bool checkEmergencyGeneral(Logger &log, EmergencyBrakes embrakes_data, Navigation nav_data,
+                                  Batteries batteries_data, Telemetry telemetry_data,
+                                  Sensors sensors_data, Motors motors_data)
+{
+  if (telemetry_data.emergency_stop_command) {
+    log.ERR("STM", "STOP command received");
+    return true;
+  } else if (nav_data.module_status == ModuleStatus::kCriticalFailure) {
+    log.ERR("STM", "Critical failure in navigation");
+    return true;
+  } else if (telemetry_data.module_status == ModuleStatus::kCriticalFailure) {
+    log.ERR("STM", "Critical failure in telemetry");
+    return true;
+  } else if (motors_data.module_status == ModuleStatus::kCriticalFailure) {
+    log.ERR("STM", "Critical failure in motors");
+    return true;
+  } else if (embrakes_data.module_status == ModuleStatus::kCriticalFailure) {
+    log.ERR("STM", "Critical failure in embrakes");
+    return true;
+  } else if (batteries_data.module_status == ModuleStatus::kCriticalFailure) {
+    log.ERR("STM", "Critical failure in batteries");
+    return true;
+  }
+  return false;
+}
+
+State *checkEmergencyStationary(Logger &log, EmergencyBrakes embrakes_data, Navigation nav_data,
+                                Batteries batteries_data, Telemetry telemetry_data,
+                                Sensors sensors_data, Motors motors_data)
+{
+  if (!checkEmergencyGeneral(log, embrakes_data, nav_data, batteries_data, telemetry_data,
+                             sensors_data, motors_data)) {
+    return NULL;
+  }
+  return FailureStopped::getInstance();
+}
+
+State *checkEmergencyMoving(Logger &log, EmergencyBrakes embrakes_data, Navigation nav_data,
+                            Batteries batteries_data, Telemetry telemetry_data,
+                            Sensors sensors_data, Motors motors_data)
+{
+  if (!checkEmergencyGeneral(log, embrakes_data, nav_data, batteries_data, telemetry_data,
+                             sensors_data, motors_data)) {
+    return NULL;
+  }
+  return FailureBraking::getInstance();
+}
```

Comment:
These don't do much, especially when you consider that the states that use them do the same if-statement all over again. Seems like an unnecessary fragmentation of the logic of those states - the code there would not be one bit more complex or longer if the states decided for themselves what the next state should be (and used `checkEmergencyGeneral()` directly).

Reply:
Those will be removed once I've changed the transition conditions to boolean.